accepted/00000-jsonb-pillar-db-storage.md

```diff
@@ -0,0 +1,219 @@
+                         # dict, descend.
+                         crd = crd[ret[i]]
+
++            # We have just one field without any key, assume returned row is already a dict
++            # aka JSON storage
++            if self.num_fields == 1:
++                crd.update(ret[0])
++
+             # If this test is true, the penultimate field is the key
+-            if self.depth == self.num_fields - 1:
++            elif self.depth == self.num_fields - 1:
+                 nk = self.num_fields-2  # Aka, self.depth-1
+                 # Should we and will we have a list at the end?
+                 if ((self.as_list and (ret[nk] in crd)) or
+```
+
+## Maintaining consistency
+
+Responsibility to provide up to date data remains, as is the case today, on the component changing the data. That means, when information about minion is changed, then relevant pillar generator must be called to refresh the pillar in database.
+However with the pillar in database and loosely (no foreign key relation) connected to minions and groups, database trigger which can remove pillar entries when minion or group is removed:
+
+```sql
+CREATE OR REPLACE FUNCTION minion_removed() RETURNS TRIGGER AS $$
+BEGIN
+  DELETE FROM suseSaltPillars WHERE target = OLD.minion_id;
+  RETURN OLD;
+END $$ LANGUAGE PLPGSQL;
+
+CREATE OR REPLACE FUNCTION group_removed() RETURNS TRIGGER AS $$
+BEGIN
+  DELETE FROM suseSaltPillars WHERE target = OLD.id::VARCHAR;
+  RETURN OLD;
+END $$ LANGUAGE PLPGSQL;
+
+CREATE TRIGGER minion_removed BEFORE DELETE ON suseMinionInfo FOR EACH ROW EXECUTE PROCEDURE minion_removed();
+
+CREATE TRIGGER group_removed BEFORE DELETE ON rhnServerGroup FOR EACH ROW EXECUTE PROCEDURE group_removed();
+```
+
+## Performance impact
+
+I did a small scale test with 100 minions and 100 pillar refresh repetitions with following results:
+
+Command:
+
+```bash
+time bash -c 'for i in `seq 1 100`;do salt "*" saltutil.pillar_refresh > /dev/null; done'
+```
+
+* both suma-minion.py and DB pillars active:
+
+```
+real    4m40,897s
+user    1m52,634s
+sys     0m10,143s
+```
+
+* only suma-minion.py pillar active:
+
+```
+real    4m37,656s
+user    1m53,964s
+sys     0m10,080s
+```
+
+
+* only DB pillar active:
+
+```
+real    4m39,619s
+user    1m54,091s
+sys     0m10,101s
+```
+
+
+I also did another small scale test with 1000 minions and 10 pillar refresh repetitions with following results:
+
+Command:
+```bash
+time bash -c 'for i in `seq 1 10`;do salt "*" saltutil.pillar_refresh > /dev/null; done'
+```
+
+* both DB pillar and `suma_minion.py` pillars active:
+
+```
+real    3m17,522s
+user    0m39,096s
+sys     0m2,260s
+```
+
+* when only `suma_minion.py` pillar was active:
+
+```
+real    3m22,705s
+user    0m40,239s
+sys     0m2,244s
+```
+
+* when only DB pillar was active:
+
+```
+real    3m21,272s
+user    0m39,661s
+sys     0m2,320s
+```
+
+In conclusion I do not see very notable impact in this isolated scenario.
```

Comment:
That's excellent. One important point: how many Postgres connections were used during these tests?

It would be a problem if the answer is "one per concurrent minion access".

Reply:
I fear that this is indeed the case. SQL external pillar itself doesn't have any caching or maintaining of single connection. Instead it for each pillar request not handled by salt pillar in-memory cache opens a new connection.
However I did not see failures due to reaching max connections. Postgres configured to 400 max, having 1001 minions.

Reply:
This is a bit strange. One possibility is that the requests were done quicker than the connections were exhausted - but still this is an important potential problems, as we know Apache and possibly Taskomatic will have problems if they starve out of connections.

Could you try again and really monitor the number of open TCP sockets, to see what is actually going on?

How difficult would it be to solve this on the Salt side? Another possibility would be to add a whole-system connection pooler, which is an option we had thought about before but is kind of yet-another-separate-effort.

Reply:
I will redo the test. I suppose that salt worker threads limits the workload here - by default we have 8 so I suppose 8 threads are handling `pillar_refresh` events and thus even though there will be new connection per minion only `worker_threads`# concurrently.

> How difficult would it be to solve this on the Salt side? Another possibility would be to add a whole-system connection pooler, which is an option we had thought about before but is kind of yet-another-separate-effort.

I am not sure, I don't have enough knowledge about implementing external pillars. I will look into other pillar modules, maybe it is already solved somewhere else.

Reply:
I redone the tests, noticed that `saltutil.refresh_pillars` do not open the connection to database, which is surprising to me, however `pillar.items` does as it ignores inmemory cache by design. So I redone all tests using `pillar.items`.

Wrt. connections I saw:
tcp: 2 - 14 (SUMA was shutdown, only salt master was running), 
`numbackends`: 2 - 7  (from `pg_stat_database`)

Hypothesis is that it is indeed limited to configured salt worker threads.

I increased salt `worker_threads` from 8 to 28 and I saw increase of tcp connections to postgres and `numbackends` as well:
tcp: 2 - 44
`numbackends`: 2 - 14

I will update RFC with these results shortly.